### PR TITLE
Martin/ios 478 re add read animation

### PR DESCRIPTION
**Pull request by MartinJibber**
This works really well I think.

- Sending: No delivery status shown. (If it fails we show the red exclamation)
- Sent: The checkmark animates in smoothly. For non-user messages, no check mark is shown. Instead we show the eye closed.
- Reading: The eye slowly opens.
- Read: The eye is open.
- Error: Exclamation mark.

https://user-images.githubusercontent.com/88675954/154776844-4b9c0a02-3621-40f6-ae59-e1c18781b3d2.mp4

### CODE REVIEW

File: Shared/Models/Message Model/Messageable.swift [hunk 1]
```diff
@@ -55,6 +56,10 @@ func ==(lhs: Messageable, rhs: Messageable) -> Bool {
 
 extension Messageable {
 
+    var streamCid: ChannelId {
+        return try! ChannelId(cid: self.conversationId)
+    }
```
Comment (MartinJibber) [inline Shared/Models/Message Model/Messageable.swift:62]: Convenient.

File: Shared/Models/Message Model/MockMessage.swift [hunk 2]
```diff
@@ -19,7 +19,11 @@ struct MockMessage: Messageable {
     var avatar: Avatar? = nil
     var status: MessageStatus = .delivered
     var context: MessageContext = .respectful
-    var hasBeenConsumedBy: [Avatar] = []
+    var hasBeenConsumedBy: [Avatar] = [SystemAvatar(userObjectId: nil,
+                                                    givenName: "system",
+                                                    familyName: "avatar",
+                                                    image: nil,
+                                                    handle: "system")]
```
Comment (MartinJibber) [inline Shared/Models/Message Model/MockMessage.swift:26]: For testing

File: Shared/Resources/Animations/alertCircle.json [hunk 3]
```diff
@@ -0,0 +1 @@
...[0]},{"t":45,"s":[100]}],"ix":7},"sk":{"a":0,"k":0,"ix":4},"sa":{"a":0,"k":0,"ix":5},"nm":"Transform"}],"nm":"cyrcle 2","np":2,"cix":2,"bm":0,"ix":3,"mn":"ADBE Vector Group","hd":false},{"ty":"gr","it":[{"ind":0,"ty":"sh","ix":1,"ks":{"a":0,"k":{"i":[[-5.523,0],[0,-5.523],[5.522,0],[0,5.522]],"o":[[5.522,0],[0,5.522],[-5.523,0],[0,-5.523]],"v":[[0,-10],[10,0],[0,10],[-10,0]],"c":true},"ix":2},"nm":"Path 1","mn":"ADBE Vector Shape - Group","hd":false},{"ty":"st","c":{"a":0,"k":[0,0,0,1],"ix":3},"o":{"a":0,"k":100,"ix":4},"w":{"a":0,"k":2,"ix":5},"lc":2,"lj":2,"bm":0,"nm":"Stroke 1","mn":"ADBE Vector Graphic - Stroke","hd":false},{"ty":"tr","p":{"a":0,"k":[12,12],"ix":2},"a":{"a":0,"k":[0,0],"ix":1},"s":{"a":1,"k":[{"i":{"x":[0.833,0.833],"y":[0.833,0.833]},"o":{"x":[0.167,0.167],"y":[0.167,0.167]},"t":0,"s":[85,85]},{"t":42,"s":[104,104]}],"ix":3},"r":{"a":0,"k":0,"ix":6},"o":{"a":1,"k":[{"i":{"x":[0.833],"y":[0.833]},"o":{"x":[0.167],"y":[0.167]},"t":0,"s":[100]},{"t":42,"s":[0]}],"ix":7},"sk":{"a":0,"k":0,"ix":4},"sa":{"a":0,"k":0,"ix":5},"nm":"Transform"}],"nm":"cyrcle","np":2,"cix":2,"bm":0,"ix":4,"mn":"ADBE Vector Group","hd":false}],"ip":0,"op":46,"st":0,"bm":0}],"markers":[]}
```
Comment (MartinJibber) [inline Shared/Resources/Animations/alertCircle.json:1]: Adding in delivery status animations.

File: iOS/_Stream Temp/ChatMessage+Extensions.swift [hunk 4]
```diff
@@ -107,6 +107,10 @@ extension Message: Messageable {
         return emotion
     }
 
+    static func message(with cid: ConversationId, messageId: MessageId) -> Message {
+        return MessageController.controller(cid, messageId: messageId).message!
+    }
```
Comment (MartinJibber) [inline iOS/_Stream Temp/ChatMessage+Extensions.swift:127]: Convenience.

File: Shared/Models/Message Model/Messageable.swift [hunk 5]
```diff
@@ -7,11 +7,13 @@
...
 
-enum MessageStatus: String {
-    case sent // Message was sent as a system message
-    case delivered // Message was successfully delivered by Twilio
-    case unknown
+enum DeliveryStatus {
+    case sending
+    case sent
+    case reading
+    case read
```
Comment (MartinJibber) [inline Shared/Models/Message Model/Messageable.swift:16]: Updating these state cases

File: iOS/UI/Swipeable Input/MessageDeliveryStatusUIViewRepresentable.swift [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+import StreamChat
+
+struct MessageDeliveryStatusUIViewRepresentable: UIViewRepresentable {
+
+    @Binding var message: Messageable?
+    @Binding var deliveryStatus: DeliveryStatus
+
+    func makeUIView(context: UIViewRepresentableContext<MessageDeliveryStatusUIViewRepresentable>) -> MessageDeliveryStatusUIView {
+        return MessageDeliveryStatusUIView()
+    }
```
Comment (MartinJibber) [inline iOS/UI/Swipeable Input/MessageDeliveryStatusUIViewRepresentable.swift:20]: Create the wrapped ui view here.

File: iOS/UI/Swipeable Input/MessageDeliveryStatusUIViewRepresentable.swift [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+    func updateUIView(_ uiView: MessageDeliveryStatusUIView, context: Context) {
+        guard let message = self.message else { return }
+
+        uiView.update(with: message, deliveryStatus: self.deliveryStatus)
+    }
```
Comment (MartinJibber) [inline iOS/UI/Swipeable Input/MessageDeliveryStatusUIViewRepresentable.swift:26]: Any updates to your data binded data will trigger this function to be called.

File: iOS/UI/Swipeable Input/MessageDeliveryStatusUIViewRepresentable.swift [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+            self.readStatusView.stop()
+            self.readStatusView.play(fromProgress: 1, toProgress: 0, loopMode: .playOnce)
+        case .read:
+            self.readStatusView.isVisible = true
+            self.readStatusView.currentProgress = 0
+        case .error:
+            self.errorStatusView.isVisible = true
+            self.errorStatusView.stop()
+            self.errorStatusView.play()
+        }
```
Comment (MartinJibber) [inline iOS/UI/Swipeable Input/MessageDeliveryStatusUIViewRepresentable.swift]: Playing animations, or snapping to animation positions based upon the message state.

File: iOS/UI/Swipeable Input/MessageDeliveryStatusUIViewRepresentable.swift [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+            self.previousStatus = deliveryStatus
+        }
+
+        self.readStatusView.isVisible = false
+        self.sendStatus.isVisible = false
+        self.errorStatusView.isVisible = false
+
+        switch deliveryStatus {
+        case .sending:
+            break
```
Comment (MartinJibber) [inline iOS/UI/Swipeable Input/MessageDeliveryStatusUIViewRepresentable.swift:87]: I don't show any UI for sending. It looks cleaner this way. If there is an issue, then an error will show.

File: iOS/_Stream Temp/ChatMessage+Extensions.swift [hunk 10]
```diff
@@ -33,7 +33,22 @@ extension Message: Messageable {
...
+                return .error
+            case .pendingSync, .syncing, .syncingFailed, .deleting:
+                break
+            }
+        }
+
+        if self.isConsumed {
+            return .read
+        }
+
```
Comment (MartinJibber) [inline iOS/_Stream Temp/ChatMessage+Extensions.swift:51]: Determining delivery state of a message.

**Review by Benjamindodgson — APPROVED** [review]

**Pull request merged**